### PR TITLE
Ledger Massive Upgrade with Major Bug Fixes

### LINKED ISSUE #1116: [Bug report] Weekly Quests keeps resetting even when user did not request for one
**Date Seen**:
01-02-2016

**Version**:

**KC3 Kai v260 (git build) ID**: jljpednjlkddlmahfoljfjlnbollampe
**KC3 Kai v260 git revision**: 3a6562c64d2aab9f72e2e5f34b4cbfa53c22b8d9
**Browser**: Chromium 64bit (currently on: 48.0.2564.97 (Developer Build 76bc56ae79e5f501eaf1ed311ff828f6daf8538a Linux))
**OS**: Archlinux 64bit (currently on: Linux: 4.3.3-3-ARCH)

**Bug Description**:
Weekly Quest keeps on resetting even though it is already new week. Selecting the weekly quests in game would make it appear on KC3 Kai panel on that screen only but would disappear when the user either goes to another quest page or exits out of quest screen.

Weekly Quest will continue resetting as noted when debugging KC3 Kai panel even when the user visits other parts of the game (e.g. "receiving" fleets from expeditions).

*This bug also happens on Google Chrome browser, not just Chromium builds. Below is a snippet output on chrome://version from Google Chrome on Linux.* 
**Google Chrome**: 48.0.2564.97 (Official Build) (64-bit)
**Revision**: 97d2b858fc7c3d480a52f4b58ed72c76fa535598-refs/branch-heads/2564@{#&#8203;631}
**OS**: Linux 
**Blink**: 537.36 (@&#8203;97d2b858fc7c3d480a52f4b58ed72c76fa535598)
**JavaScript**: V8 4.8.271.17
**Flash**: 20.0.0.286
**User Agent**: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/48.0.2564.97 Safari/537.36
**Command Line**: /opt/google/chrome/google-chrome --flag-switches-begin --flag-switches-end

Please refer to the screenshots provided below.

**Severity**:
Moderate - Major

**Steps to Reproduce**:
1. Start KC3 Kai and KanColle as usual.
2. Go into Quests inside the game
3. Go and select weekly expeditions. In my case I have selected weekly expedition quests such as complete 30 expeditions as well as complete Tokyo express expedition once.
4. Go either into another Quests page or exit out of Quests section in-game.
5. Notice the KC3 Kai panel no longer displays the weekly quests that you previously selected. 

*Open KC3 Kai debug console*
1. Go into Quests page in-game.
2. Click on any of the quests pages or exit out of quests pages in-game.
3. Notice how the debug console keeps informing that the weekly quests has been reset.

**Actual Behavior**:
- Weekly quests keeps on resetting

**Expected Behavior**:
- Weekly quest should only be reset once, not for every time the user visits other in-game pages.

**Troubleshooting/Testing Steps Attempted**:
- Pulled the latest version of KC3 Kai development version and the issue persists - was previously on 8e2b19ca8c6caab13bae2087af6839cd32eb31d3 when the bug first surfaced.
- Tried pressing on reset daily/weekly/monthly/everything quests button inside the strategy room just once in a bid to "un-stick" quests reset. Bug still is present.  

**Workaround**:
- Perhaps ````git rebase```` on older git commits would resolve the issue but I have a pressing issue with Chromium automatically resetting my cache for Kancolle game. :sweat: 

**Attachments**:
screenshot_20160201_100236

### PR DESCRIPTION
**Pull request by ReiFan49**
basically this one applies some fixes to make sure the ledger looks better.
preparing the sortie stat despair feature along the ledger due to it's dependency.

- Fixes #1116
- Waits for #1170

### CODE REVIEW

**Comment by dragonjet** [comment]
is this alright to merge? or is there any other pending dependency something?
I'm thinking we can merge this since we're already into the ledger in master

**Comment by ReiFan49** [comment]
hmmm to mark a large change ? a bit confused but still waiting for the final filtering feature that will be done on ledger,

[Chat Explanation](https://gitter.im/KC3Kai/Public?at=56a4bff76b6468374a0930fc) about the feature

**Comment by ReiFan49** [comment]
So basically i'm preparing a major patch,
Having issues with memory management lately lol,
Gotta change a bit how does the ledger handles the data and so on.

**Comment by dragonjet** [comment]
Is this alright to merge? XD
If this includes the fix to https://github.com/KC3Kai/KC3Kai/issues/1116, then merging for 26.0 is required
But what about the memory management? Is it not advisable to merge before that issue is dealt with?

**Comment by dragonjet** [comment]
Is this alright to merge? XD
If this includes the fix to https://github.com/KC3Kai/KC3Kai/issues/1116, then merging for 26.0 is required
But what about the memory management? Is it not advisable to merge before that issue is dealt with?

**Comment by ReiFan49** [comment]
the memory issue might happen later (as people expected to have low sample size around ~1500 ledger queries/actions during an event), people might complaint the processing speed due to large sample that they have (my condition is 4000+ data, which is not advised to perform database query especially whole duration).
right now i only use cold-caching style (with unlimited size) that calculated with `SortieLedgerCount x LedgerRecordedEventMaps` (expect the result to be around 7 digits of number on later state after 2~3 months data collection) length of array.

caching the ledger data improves the speed greatly while changing the date on natural way (whole <-> dailies/weeklies/monthlies/yearlies).

**Comment by ReiFan49** [comment]
uh oh
image
i didn't realize it... :sweat_smile: (attempt to push -- before 12PM today, just checked... after 7:30PM today...)

**Comment by ReiFan49** [comment]
ok, memory problem is fixed,
- Load 5000 samples from database: ~2 seconds
- Process Sortie Ledger from 5000 samples: ~1 second
- Process Event Ledger (from Map Start to Finish and such) from 5000 samples: ~6 seconds

terrifying performance boost that increases up to 97% of the base speed.

**Comment by dragonjet** [comment]
:+1: hopefully we can get @KC3Kai/testers checking out to this branch for verification :sweat_smile:

**Comment by dragonjet** [comment]
can you pull from master into `dev-lodger`. I'm afraid I'll break something if I do the conflict resolution.

**Comment by dragonjet** [comment]
marking `needs-corrections` because of the travis checks
once travis passed, I'll merge this into master
and move on to #1155

**Comment by sinsinpub** [comment]
this one has been merged into #1155 and we corrected it there

**Comment by LostLogia4** [comment]
really? :open_mouth: I couldn't see any new commit on either PR.

**Comment by sinsinpub** [comment]
I think so. because i found the git log that `dev-lodger` merged into `dev-api`, and I merged `dev-api` into `dev-more-i18n`
may needed @ReiFan49 to confirm it :smile:

File: src/library/modules/Kcsapi.js [hunk 1]
```diff
@@ -1525,73 +1573,57 @@ Previously known as "Reactor"
...
-				}
-			}
-
-			// Vague quest that clears with no rank requirement
-			qLog(216).increment(); // Bd2: Defeat the flagship of an enemy fleet
-			
-			// If node is a boss
-			if( KC3SortieManager.currentNode().isBoss() ){
-				qLog(214).increment(1); // Bw1: 2nd requirement: Encounter 24 bosses (index:1)
-			}
```
Comment (LostLogia4) [inline src/library/modules/Kcsapi.js]: @dragonjet here's the ecma6 code.

**Comment by dragonjet** [comment]
oh my what have I done :fearful:

**Comment by LostLogia4** [comment]
@dragonjet  what happened?

File: src/library/modules/Kcsapi.js [hunk 2]
```diff
@@ -1525,73 +1573,57 @@ Previously known as "Reactor"
...
+					return (
+						(!questData[2] || KC3SortieManager.isSortieAt.apply(KC3SortieManager,questData[2])) && /* Is sortie at */
+						(!questData[3] || KC3SortieManager.currentNode().isBoss())                       && /* Is on boss node */
+						true;
+					);
+				})
+				// Quest data increment on given index
+				.forEach(function(questData){
+					qLog(questData[0]).increment(questData[1]); // note that every sortie quests are logged off
+				});
```
Comment (LostLogia4) [inline src/library/modules/Kcsapi.js]: @dragonjet here's the ecma6 code.

File: src/library/modules/Kcsapi.js [hunk 3]
```diff
@@ -1525,73 +1573,57 @@ Previously known as "Reactor"
...
+					[261,0,[1,5], true], // Bw10: Sortie to [W1-5] and A-rank+ the boss node 3 times
+					[265,0,[1,5], true]  // Bm5: Deploy a fleet to [W1-5] and A-rank+ the boss node 10 times
+				],
+				[ /* S RANK */
+					[214,3,false,false], // Bw1: 4th requirement: 6 S ranks (index:3)
+					[243,0,[5,2], true], // Bw9: Sortie to [W5-2] and S-rank the boss node 2 times
+					[256,0,[6,1], true]  // Bm2: Deploy to [W6-1] and obtain an S-rank the boss node 3 times
+				],
+				[ /* KANZEN */ ],
+			].reverse().splice(rankPt)
```
Comment (LostLogia4) [inline src/library/modules/Kcsapi.js]: @dragonjet here's the ecma6 code.

**Comment by dragonjet** [comment]
@LostLogia4 nah I was just surprised this was also marked as "merged" when I did the i18n merge. I did not push the button on this thread, but on #1155. Didn't know github detects it automatically

**Comment by sinsinpub** [comment]
I have reverted this function to the old revision of master in my branch, don't worry
but I've not verified on ES6 at all. if any other found, we could check it out :sweat_smile:

File: src/library/modules/Kcsapi.js (hunk 3 above)
Comment (ReiFan49) [inline src/library/modules/Kcsapi.js]: actually the ECMA6 on that place is **already** removed,

**Pull request merged**